### ISSUE TITLE
Question about AP definition

### ISSUE THREAD

**Issue by 0phoff** [cocodataset/cocoapi#268]
I was wondering what the COCO api uses as definition for the AP metric (for object detection).
Seeing as this project is used by a lot of researchers to evaluate the performance of their object detection models, I would like to have more details about the chosen definition.  
As I understand it, there are a few different ways to compute AP:

  1. You can make the Precision monotonically decreasing, by taking the maximal precision of every  
`R >= Rn` for every point n on the curve. This removes the sawtooth-jiggles from the curve and is done in the [Pascal VOC metric](http://host.robots.ox.ac.uk/pascal/VOC/voc2012/htmldoc/devkit_doc.html#sec:ap).
  2. There are a few ways to perform the actual AP computation:
    - 11-point interpolated AP: compute the PR-values at [0, 0.1, ..., 1] and compute the AP as the arithmetic mean of these values (works because delta-r is constant).
    - Following the definition of AP according to [scikit-learn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.average_precision_score.html#sklearn.metrics.average_precision_score) and [wikipedia](https://en.wikipedia.org/w/index.php?title=Information_retrieval&oldid=793358396#Average_precision) (not the best sources, but all I could seem to find in terms of proper definition).
This means taking the sum of `Pn * (Rn - Rn-1)` and is like approximating the Area under the curve with rectangles at each (P,R) point.
    - Approximating the AP by computing the Area under the curve (AUC). This is usually done by using the trapezium rule.

I would want to know this to be able to compare my results with other papers which use your framework, but I could not seem to find any explanation on the COCO website and the (python) code is quite difficult to get into. It could be nice to have this properly explained somewhere? (Or I completely missed it, in which case I'm sorry for opening this issue)

The differences between the different methods might not be that big, bug they are noticeable (mainly that the the first 2 methods are usually a small overestimation of the proper AUC).  
Using the AUC seems the most correct metric to me, but then you would have called it AUC and not AP I guess? So I am not sure what you guys are using...

Thanks a lot in advance!

**Comment by harrisonford** [comment]
I would like to know more about the AP metric too, I wonder if "AP-head" is implemented internally in the cocotools?

**Comment by Menooker** [comment]
@0phoff @harrisonford Hi, I have found in the COCO Python implementation that it uses 11-point  interpolated AP. You can follow the link:
https://github.com/cocodataset/cocoapi/blob/636becdc73d54283b3aac6d4ec363cffbb6f9b20/PythonAPI/pycocotools/cocoeval.py#L507

The variable `recThrs` should be "recall threshold" and it should be [0, 0.1,... 1.0] in the code

**Comment by thho** [comment]
@Menooker not sure how you interpret line 507 in 636becd as a 11 point interpolation,
https://github.com/cocodataset/cocoapi/blob/636becdc73d54283b3aac6d4ec363cffbb6f9b20/PythonAPI/pycocotools/cocoeval.py#L26
providing the hint for a 101 point interpolation over the recall curve.

**Comment by Menooker** [comment]
> @Menooker not sure how you interpret line 507 in [636becd](https://github.com/cocodataset/cocoapi/commit/636becdc73d54283b3aac6d4ec363cffbb6f9b20) as a 11 point interpolation,
> https://github.com/cocodataset/cocoapi/blob/636becdc73d54283b3aac6d4ec363cffbb6f9b20/PythonAPI/pycocotools/cocoeval.py#L26
> 
> 
> providing the hint for a 101 point interpolation over the recall curve.

Oh... It's a typo. It should be 101-point as you pointed out, since `np.round((1.00 - .0) / .01) + 1` = 101

**Comment by 0phoff** [comment]
So they use yet another way...  
Why is it so hard to have a single definition that is used by everyone ?

Any idea why one would use 101-point interpolation ?

I guess with that many points the differences between this AP and the scikit one would be minimal, but this one will probably ignore strong spikes more ?  

I think I will just implement the scikit version in my own library, as that is the one with the definition I encounter the most when searching through the webs about AP.
